include/seqan/seeds/seeds_global_chaining.h

```diff
@@ -232,51 +242,41 @@ chainSeedsGlobally(
-                    // Roughly then, there is no chain ending in a seed below the seed_k, that has a lower score
-                    // than the chain ending in seed_k. Thus the last value in `intermediateSolutions` will
-                    // always point to the optimal chain.
-                    TIntermediateSolutionsIterator itDel = intermediateSolutions.upper_bound(referenceSolution);
-                    TIntermediateSolutionsIterator itDelEnd = intermediateSolutions.end();
-                    while (itDel != itDelEnd)
-                    {
-                        TIntermediateSolutionsIterator ptr = itDel;
-                        ++itDel;
-                        if (qualityOfChainEndingIn[it_k->i3] > ptr->i2)
-                        {
-                            // std::cout << "  ERASE (" << ptr->i1 << ", " << ptr->i2 << ", " << ptr->i3 << ")" << std::endl;
-                            intermediateSolutions.erase(ptr);
-                        }
-                    }
-                }
-            } // ... otherwise, add a triple for k in L if either L is empty or the last triple in
-              // L has a lower score than the chain ending in k.
-            else if (intermediateSolutions.empty() || (--it_j)->i2 < qualityOfChainEndingIn[it_k->i3])
+            TIntermediateSolution sol(endPositionV(seed_k), qualityOfChainEndingIn[it_k->i3], it_k->i3);
+            // std::cout << "  INSERT (" << sol.i1 << ", " << sol.i2 << ", " << sol.i3 << ")" << __LINE__  << std::endl;
+            intermediateSolutions.insert(sol);
+
+            // Delete all intermediate solutions where end1 >= end1 of k and have a lower score than k
+            // to ensure that the invariant of V(j) >= V(j'), with j' <= j holds.
+            // Roughly then, there is no chain ending in a seed below the seed_k, that has a lower score
+            // than the chain ending in seed_k. Thus the last value in `intermediateSolutions` will
+            // always point to the optimal chain.
+            TIntermediateSolutionsIterator itDel = intermediateSolutions.upper_bound(referenceSolution);
+            TIntermediateSolutionsIterator itDelEnd = intermediateSolutions.end();
+            while (itDel != itDelEnd)
             {
-                // None found.  Insert a new triple for seed k.
-                TIntermediateSolution sol(endPositionV(seed_k), qualityOfChainEndingIn[it_k->i3], it_k->i3);
-                // std::cout << "  INSERT (" << sol.i1 << ", " << sol.i2 << ", " << sol.i3 << ") " << __LINE__ << std::endl;
-                intermediateSolutions.insert(sol);
+                TIntermediateSolutionsIterator ptr = itDel;
+                ++itDel;
+                if (qualityOfChainEndingIn[it_k->i3] > ptr->i2)
+                {
+                    // std::cout << "  ERASE (" << ptr->i1 << ", " << ptr->i2 << ", " << ptr->i3 << ")" << std::endl;
+                    intermediateSolutions.erase(ptr);
+                }
```

Comment:
you must break here if this condition is false, because the condition will stay false for all following iterations

otherwise you might not achieve the `r log r` time